### PR TITLE
Fix/public suffix list tests

### PR DESCRIPTION
**Pull request by massdosage**
Rough initial draft with all tests working.

### CODE REVIEW

File: httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java [hunk 1]
```diff
@@ -149,7 +156,9 @@ public String getDomainRoot(final String domain, final DomainType expectedType)
             }
             final DomainType domainRule = findEntry(rules, key);
             if (match(domainRule, expectedType)) {
-                if (domainRule == DomainType.PRIVATE) {
+                // HUMAN code change - PSL algorithm doesn't have any rules changing the result based on "domain type"
```
Comment (braedon) [inline httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java]: I presume we don't want to say `HUMAN code change` in the upstream version.

Comment (braedon) [reply httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java]: Probably need to talk explicitly about `legacyMode` in this comment.

Comment (braedon) [reply httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java]: (ditto for other `HUMAN code change` comments below)

File: httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java [hunk 2]
```diff
@@ -158,21 +167,42 @@ public String getDomainRoot(final String domain, final DomainType expectedType)
             final int nextdot = segment.indexOf('.');
             final String nextSegment = nextdot != -1 ? segment.substring(nextdot + 1) : null;
 
-            if (nextSegment != null) {
-                final DomainType wildcardDomainRule = findEntry(rules, "*." + IDN.toUnicode(nextSegment));
+            if (!legacyMode) {
```
Comment (braedon) [inline httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java]: I'm a bit torn on this, but it might be clearer if we switch the order, and use `if (legacyMode)` here.

File: httpclient5/src/test/java/org/apache/hc/client5/http/impl/cookie/TestPublicSuffixListParser.java [hunk 3]
```diff
@@ -111,7 +111,7 @@ void testParseLocal() {
 
         cookie.setDomain(".blah");
         cookie.setAttribute(Cookie.DOMAIN_ATTR, ".blah");
-        Assertions.assertTrue(filter.match(cookie, new CookieOrigin("somehost.blah", 80, "/stuff", false)));
+        Assertions.assertTrue(filter.match(cookie, new CookieOrigin("somehost.blah", 80, "/stuff", false), true));
```
Comment (braedon) [inline httpclient5/src/test/java/org/apache/hc/client5/http/impl/cookie/TestPublicSuffixListParser.java]: We shouldn't need to use legacyMode here, as the test PSL file doesn't contain any `DomainType.PRIVATE` entries (or `DomainType.ICANN` entries for that matter - it seems to be parsed with a separate parser that doesn't handle the sections and sets everything to `DomainType.UNKNOWN`?).

Has the behavior changed because of our proper implementation of the default wildcard rule perhaps?

File: httpclient5/src/test/java/org/apache/hc/client5/http/psl/TestPublicSuffixMatcher.java [hunk 4]
```diff
@@ -55,10 +61,10 @@ void setUp() throws Exception {
...
     void testGetDomainRootAnyType() {
         // Private
-        Assertions.assertEquals("xx", matcher.getDomainRoot("example.XX"));
-        Assertions.assertEquals("xx", matcher.getDomainRoot("www.example.XX"));
-        Assertions.assertEquals("xx", matcher.getDomainRoot("www.blah.blah.example.XX"));
-        Assertions.assertEquals("appspot.com", matcher.getDomainRoot("example.appspot.com"));
+        Assertions.assertEquals("xx", matcher.getDomainRoot("example.XX", true));
+        Assertions.assertEquals("xx", matcher.getDomainRoot("www.example.XX", true));
+        Assertions.assertEquals("xx", matcher.getDomainRoot("www.blah.blah.example.XX", true));
+        Assertions.assertEquals("appspot.com", matcher.getDomainRoot("example.appspot.com", true));
```
Comment (braedon) [inline httpclient5/src/test/java/org/apache/hc/client5/http/psl/TestPublicSuffixMatcher.java]: It'd be good to copy all the legacy mode tests to a separate section (or sections) and (attempt to) explain what behavior it's "supposed" to have.

Then we can change these versions to use non-legacy mode, and assert the correct result.

File: httpclient5/src/test/java/org/apache/hc/client5/http/impl/cookie/TestPublicSuffixListParser.java (hunk 3 above)
Comment (braedon) [reply httpclient5/src/test/java/org/apache/hc/client5/http/impl/cookie/TestPublicSuffixListParser.java]: OK, I think I worked it out. (Explanation mainly for my benefit)

The filter is trying to prevent (i.e. filter out by returning false for) cookies being set on public suffixes (i.e. eTLDs, not eTLD+1s), to prevent registered domains (eTLD+1s) from setting cookies for all domains (and their subdomains) under a given public suffix/eTLD.

The `PublicSuffixMatcher.matches()` method tries to return true if the input domain is a public suffix. It does this by checking `PublicSuffixMatcher.getDomainRoot()`, and if it returns `null`, assuming the input was a public suffix, as public suffixes don't have root domains. (There are also other cases where this implementation returns `null` for inputs that aren't public suffixes, but I don't _think_ any are pertinent here...)

Under the legacy `PublicSuffixMatcher.getDomainRoot()` implementation, if you pass it a single-label "domain" (a TLD) that isn't explicitly in the PSL, it'll return it. So if you pass it `blah`, it'll return `blah`, which `PublicSuffixMatcher.matches()` interprets as meaning it's _not_ a public suffix, so this filter returns `true` and doesn't filter out the cookie, allowing `somehost.blah` to set a cookie on it for all subdomains of `blah`.

However, this is incorrect. The PSL algorithm states the "prevailing rule" if no rule matches is `*`, which means the public suffix of `blah` is `blah`, so `PublicSuffixMatcher.getDomainRoot()` should return `null` (and does under the new rules). This means `blah` is a public suffix, so the cookie _shouldn't_ be set, as this would allow `somehost.blah` to set a cookie for all other domains registered under `blah`.

tl;dr
IMO the test is wrong, and the legacy functionality is dangerous. It allows any registered domain (or subdomain) to set a cookie on any TLD that's not explicitly in the PSL.

I think we should change this test, and drop the unnecessary/dangerous `legacyMode` flag from `PublicSuffixDomainFilter.match()`, and probably from `PublicSuffixMatcher.matches()` as well.

File: httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java (hunk 1 above)
Comment (massdosage) [reply httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java]: Yeah, I'll remove all those, were in our other fork to be clear where the changes were.

File: httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java (hunk 2 above)
Comment (massdosage) [reply httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java]: I'll switch it and you can take a look and see which you prefer ;)

File: httpclient5/src/test/java/org/apache/hc/client5/http/impl/cookie/TestPublicSuffixListParser.java (hunk 3 above)
Comment (massdosage) [reply httpclient5/src/test/java/org/apache/hc/client5/http/impl/cookie/TestPublicSuffixListParser.java]: We can go through this together with the debugger if you'd like but we hit this return statement in the matcher if we don't go for legacy mode in this case:
```
                if (nextSegment == null && (expectedType == null || expectedType == DomainType.UNKNOWN)) {
                    return result;
                }
```

File: httpclient5/src/test/java/org/apache/hc/client5/http/psl/TestPublicSuffixMatcher.java (hunk 4 above)
Comment (massdosage) [reply httpclient5/src/test/java/org/apache/hc/client5/http/psl/TestPublicSuffixMatcher.java]: Sure, I'll make a version of each of these methods that does that.

Comment (massdosage) [reply httpclient5/src/test/java/org/apache/hc/client5/http/psl/TestPublicSuffixMatcher.java]: I'll push this in a bit, it's a good idea as this shows the difference between legacy and new behaviour and might show up some edge cases we haven't covered...

File: httpclient5/src/test/java/org/apache/hc/client5/http/impl/cookie/TestPublicSuffixListParser.java (hunk 3 above)
Comment (braedon) [reply httpclient5/src/test/java/org/apache/hc/client5/http/impl/cookie/TestPublicSuffixListParser.java]: Sorry, my second comment was sitting there in "pending" state - I thought I'd sent it...

tl;dr: When the input cookie domain is `.blah`  I think this code (in non-legacy mode) will return `null`, which will be interpreted as `blah` being a public suffix (which under the default `*` rule, it is), and the cookie filter will return false. IMO that's correct, and the test is wrong - it should assert `false`.

Comment (massdosage) [reply httpclient5/src/test/java/org/apache/hc/client5/http/impl/cookie/TestPublicSuffixListParser.java]: OK, so let me revert the usage of `legacyMode` for this thing and change this test to `AssertFalse` for now.

File: httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java [hunk 5]
```diff
@@ -118,20 +118,47 @@ private static boolean match(final DomainType domainType, final DomainType expec
...
+     * @since 5.4
+     */
+    public String getDomainRoot(final String domain, final boolean legacyMode) {
+        return getDomainRoot(domain, null, legacyMode);
+    }
+
+    /**
+     * Returns registrable part of the domain for the given domain name or {@code null}
+     * if given domain represents a public suffix. For backwards-compatability reasons this uses
+     * "legacy mode" to retain previous behaviour.
```
Comment (braedon) [inline httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java]: This constructor currently sets `legacyMode` to `false`

File: httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java [hunk 6]
```diff
@@ -149,7 +176,9 @@ public String getDomainRoot(final String domain, final DomainType expectedType)
             }
             final DomainType domainRule = findEntry(rules, key);
             if (match(domainRule, expectedType)) {
-                if (domainRule == DomainType.PRIVATE) {
+                // PSL algorithm doesn't have any rules changing the result based on "domain type"
+                // see https://github.com/publicsuffix/list/wiki/Format#formal-algorithm
+                if (legacyMode && domainRule == DomainType.PRIVATE) {
```
Comment (braedon) [inline httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java]:
```suggestion
                // Prior to version 5.4 the result for "private" rules was different. However, the
                // PSL algorithm doesn't have any rules changing the result based on "domain type"
                // see https://github.com/publicsuffix/list/wiki/Format#formal-algorithm
                if (legacyMode && domainRule == DomainType.PRIVATE) {
```

File: httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java [hunk 7]
```diff
@@ -158,21 +187,39 @@ public String getDomainRoot(final String domain, final DomainType expectedType)
...
+                    if (match(wildcardDomainRule, expectedType)) {
+                        if (wildcardDomainRule == DomainType.PRIVATE) {
+                            return segment;
+                        }
+                        return result;
                     }
+                }
+            } else {
+                // look for wildcard entries and don't change the result based on "domain type"
+                final String wildcardKey = (nextSegment == null) ? "*" : "*." + IDN.toUnicode(nextSegment);
```
Comment (braedon) [inline httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java]:
```suggestion
                // look for wildcard entries
                final String wildcardKey = (nextSegment == null) ? "*" : "*." + IDN.toUnicode(nextSegment);
```

File: httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java [hunk 8]
```diff
@@ -158,21 +187,39 @@ public String getDomainRoot(final String domain, final DomainType expectedType)
...
-                    if (wildcardDomainRule == DomainType.PRIVATE) {
-                        return segment;
+            if (legacyMode) {
+                if (nextSegment != null) {
+                    final DomainType wildcardDomainRule = findEntry(rules, "*." + IDN.toUnicode(nextSegment));
+                    if (match(wildcardDomainRule, expectedType)) {
+                        if (wildcardDomainRule == DomainType.PRIVATE) {
+                            return segment;
+                        }
+                        return result;
```
Comment (braedon) [inline httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java]:
```suggestion
                    // look for wildcard entries and change the result based on "domain type"
                    // to match legacy behavior prior to version 5.4.
                    final DomainType wildcardDomainRule = findEntry(rules, "*." + IDN.toUnicode(nextSegment));
                    if (match(wildcardDomainRule, expectedType)) {
                        if (wildcardDomainRule == DomainType.PRIVATE) {
                            return segment;
                        }
                        return result;
```

File: httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java (hunk 5 above)
Comment (massdosage) [reply httpclient5/src/main/java/org/apache/hc/client5/http/psl/PublicSuffixMatcher.java]: Yes, so we actually default to the "correct" behaviour. I'll update the comment here to reflect that.
